v2/tools/generator/internal/codegen/storage/group_conversion_graph_builder.go

```diff
@@ -33,57 +41,109 @@ func (b *GroupConversionGraphBuilder) Add(refs ...astmodel.PackageReference) {
-			links[sr.Local()] = sr
-			// Append to our list of all storage references
-			storageReferences = append(storageReferences, sr)
+	toProcess := b.references.AsSlice()
+	astmodel.SortPackageReferencesByPathAndVersion(toProcess)
+
+	for _, s := range stages {
+		s(toProcess)
+		toProcess = b.withoutLinkedReferences(toProcess)
+	}
+
+	// Expect to have only the hub reference left
+	if len(toProcess) != 1 {
+		return nil, errors.Errorf(
+			"expected to have linked all references in group %q, but have %d left",
+			b.group,
+			len(toProcess))
+	}
+
+	result := &GroupConversionGraph{
+		group: b.group,
+		links: b.links,
+	}
+
+	return result, nil
+}
+
+// compatibilityReferencesConvertForward links any compatibility references forward to the following version
+func (b *GroupConversionGraphBuilder) compatibilityReferencesConvertForward(refs []astmodel.PackageReference) {
+	for i, ref := range refs {
+		if !b.isCompatibilityPackage(ref) {
+			continue
 		}
+
+		// Safe to use i+1 because compatibility references will always be preceded by a reference to the original
+		// storage package reference
+		b.links[ref] = refs[i+1]
 	}
+}
 
-	// Create links between the storage versions
-	astmodel.SortPackageReferencesByPathAndVersion(storageReferences)
+// apiReferencesConvertToStorage links each API type to the associated storage package
+func (b *GroupConversionGraphBuilder) apiReferencesConvertToStorage(refs []astmodel.PackageReference) {
+	for _, ref := range refs {
+		if s, ok := ref.(astmodel.StoragePackageReference); ok {
+			b.links[s.Local()] = s
+		}
+	}
+}
 
-	// For each Preview API version, the link goes from the associated storage variant to the immediately prior
-	// storage variant, no matter whether it's preview or GA.
-	for i, ref := range storageReferences {
+// previewReferencesConvertBackward links each preview version to the immediately prior version, no matter whether it's
+// preview or GA.
+func (b *GroupConversionGraphBuilder) previewReferencesConvertBackward(refs []astmodel.PackageReference) {
+	for i, ref := range refs {
 		if i == 0 || !ref.IsPreview() {
 			continue
 		}
 
-		links[storageReferences[i]] = storageReferences[i-1]
+		b.links[ref] = refs[i-1]
 	}
+}
 
-	// For each GA (non-Preview) API version, the link goes from the associated storage variant to the next GA storage
-	// variant, skipping any preview releases in between (if any)
-	var gaRelease astmodel.PackageReference
-	for index, ref := range storageReferences {
-		if index == 0 {
-			// Always treat the very earliest version as GA
-			gaRelease = storageReferences[0]
+// nonPreviewReferencesConvertForward links each version with the immediately following version.
+// By the time we run this stage, we should only have non-preview (aka GA) releases left
+func (b *GroupConversionGraphBuilder) nonPreviewReferencesConvertForward(refs []astmodel.PackageReference) {
+	for i, ref := range refs {
+		if i == 0 {
```

Comment:
What's special about `i==0` here? Maybe leave a comment explaining?

Reply:
Reworked this method for clarity; included a comment too.